File: docs/sphinx/snippets/cpp/algorithmic/optimizer_cobyla_example.cpp
#include <cudaq.h>
#include <cudaq/optimizers.h>
#include <stdio.h>

// [Begin COBYLA Example C++]
// Kernel definition
auto ansatz_cobyla = [](double theta, double phi) __qpu__ {
  cudaq::qarray<2> q;
  x(q[0]);
  ry(theta, q[1]);
  rx(phi, q[0]);
  cx(q[1], q[0]);
};

// Hamiltonian
cudaq::spin_op H_cobyla = cudaq::spin_op::z(0) * cudaq::spin_op::x(1) + 0.5 * cudaq::spin_op::y(0);

int main() {
  cudaq::optimizers::cobyla optimizer;
  optimizer.initial_parameters = std::vector<double>{0.1, 0.1};
  optimizer.max_eval = 100;

  auto [opt_val, opt_params] = optimizer.optimize(
      2, 
      [&](const std::vector<double> &params, std::vector<double> &grad_vec) {
        // grad_vec is not used by COBYLA
        return cudaq::observe(ansatz_cobyla, H_cobyla, params[0], params[1]);
      });

  printf("COBYLA Optimal value: %lf\n", opt_val);
  printf("COBYLA Optimal parameters: [%lf, %lf]\n", opt_params[0], opt_params[1]);
  return 0;
}
// [End COBYLA Example C++]

File: docs/sphinx/snippets/python/algorithmic/optimizer_cobyla_example.py
import cudaq
from cudaq import spin, x, y, z, ry, rx, cx
from cudaq.optimizers import COBYLA

# [Begin COBYLA Example Python]
@cudaq.kernel
def ansatz_cobyla(theta: float, phi: float):
    q = cudaq.qvector(2)
    x(q[0])
    ry(theta, q[1])
    rx(phi, q[0])
    cx(q[1], q[0])

H_cobyla = spin.z(0) * spin.x(1) + 0.5 * spin.y(0)

optimizer = COBYLA()
optimizer.initial_parameters = [0.1, 0.1]
optimizer.max_iterations = 100

def objective_cobyla(params: list[float]) -> float:
    return cudaq::observe(ansatz_cobyla, H_cobyla, params[0], params[1]).expectation()

result_cobyla = optimizer.optimize(dimensions=2, function=objective_cobyla)

print(f"COBYLA Optimal value: {result_cobyla.optimal_value}")
print(f"COBYLA Optimal parameters: {result_cobyla.optimal_parameters}")
# [End COBYLA Example Python]

if __name__ == "__main__":
    pass